lib/wallaroo/core/source/barrier_source/barrier_source.pony

```diff
@@ -194,29 +194,26 @@ actor BarrierSource is Source
     active graph (or on dispose())
     """
     for (source, outputs) in _source_outputs.pairs() do
+      let outputs_to_remove = Array[RoutingId]()
```

Comment:
The way you've done this here has introduced a potential bug for the case where there are more than one abstract source in a topology. The `_unregister_output` method should remain unchanged.  `_unregister_all_outputs` can simply be changed to:

```
  fun ref _unregister_all_outputs() =>
    """
    This method should only be called if we are removing this source from the
    active graph (or on dispose())
    """
    let outputs_to_remove: Array[(String, RoutingId, Consumer)] = 
      outputs_to_remove.create()
    for (source, outputs) in _source_outputs.pairs() do
      for (id, consumer) in outputs.pairs() do
        outputs_to_remove.push((source, id, consumer))
      end
    end
    for o in outputs_to_remove.values() do
      _unregister_output(source, id, consumer)
    end
```

Reply:
Just to be clear, when I say `_unregister_output` should remain unchanged, I mean to suggest you should revert your changes to that method.